0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+
+=== What is the scope of a single communication protocol?
+
+Knowing that we likely want more than one communication protocol, the question is, what is the scope of one of those libp2p protocols?
+
+In order to not fall into the same trap as with the original communication protocol, we can just opt to not design libp2p protocols for things that we don't (yet) know.
+The result of this idea is that we need (at least) a single libp2p protocol for each cryptographic protocol.
+As elaborated in <<Protocols in COMIT>>, this number is currently 4.
+
+There is a downside to this:
+Duplication.
+
+Designing a completely independent libp2p protocol for each swap would mean having very similar messages with different kinds of data types.
+
+=== The oneshot substream pattern
+
+A pattern common within libp2p is the 'oneshot substream' pattern.
+It refers to the idea of defining protocols that only have a single message.
+Examples include `/ipfs/id/1.0.0/` and `/ipfs/id/push/1.0.0`: https://github.com/libp2p/specs/tree/master/identify
+
+The advantages of using single-message protocols are:
+
+* short-lived:
+Open a substream, send the message, close it again.
+No need to wait for the other party to reply.
+* very focused:
+These protocols are small and focused, avoiding the trap of over-engineering them.
+* simpler client code:
+If there is only one possible message that can be sent, its shape can be hardcoded by the client.
+Software that uses libp2p has to cater for several protocols being active concurrently anyway.
+Using the protocol layer as the level with which messages vary makes many things a lot simpler.
+In other words, no need to have `type` fields inside the message, the protocol already identifies the type.
+* cheap:
+Opening substreams is very cheap compared to network connections.
+While there is still some inefficiency in terms of opening substreams, a more flexible design seems to be desirable over some added latency.
+Plus, https://github.com/libp2p/specs/tree/master/identify[work is underway] to make this more efficient.
+
+=== Communication requirements
+
+In order to design the ideal communication protocol, we have to identify which data we want to transfer and why.
+This section aims to achieve that.
+
+==== Verifying the details of the swap
+
+One of the primary requirements for the communication protocol is for both parties to verify that they talk about the same swap.
+This is crucial in avoiding unnecessary transactions, as long as both parties are willing to actually swap.
+
+==== Sending over protocol-specific data
+
+For HTLC-based protocols like HAN, HErc20 and HALight, Alice generates a secret.
+The communication protocol must provide a way for Alice to share the hash of this secret with Bob.
+We also have to take into account that in the future, we will have swap protocols that depend on other cryptographic elements like signatures.
+
+==== Sharing identities
```

Comment:
I am not sure to understand the difference with the previous section as identity are protocol-specific data.

Reply:
I agree this section needs some explanation why it is separate.

From a discussion I had with @thomaseizinger before he left. Yes, in the context of atomic swaps identities are protocol-specific parameters. But the identities are actually independent of the swap. They exist on the ledger and are not purely created for the swap (like e.g. the secret). Thus we could handle them separate. A protocol to exchange e.g. a Bitcoin identity could be used for all kinds of use cases, not only atomic swaps.

Reply:
Let's add more info there then please.

Reply:
Done!